src/main/scala/com/microsoft/hyperspace/index/rules/ApplyHyperspace.scala

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+import com.microsoft.hyperspace.{ActiveSparkSession, Hyperspace}
+import com.microsoft.hyperspace.actions.Constants
+import com.microsoft.hyperspace.index.IndexLogEntry
+import com.microsoft.hyperspace.index.rules.ApplyHyperspace.PlanToIndexesMap
+import com.microsoft.hyperspace.telemetry.HyperspaceEventLogging
+
+/**
+ * Collect candidate indexes for each relation.
+ */
+object CandidateIndexCollector extends ActiveSparkSession {
+  // TODO: ColumnSchemaFilter :: FileSignatureFilter :: Nil
+  private val sourceFilters: Seq[SourcePlanIndexFilter] = Nil
+
+  private def initializePlanToIndexes(
+      plan: LogicalPlan,
+      indexes: Seq[IndexLogEntry]): PlanToIndexesMap = {
+    val provider = Hyperspace.getContext(spark).sourceProviderManager
+    plan.collect {
+      case l: LeafNode if provider.isSupportedRelation(l) =>
+        (l.asInstanceOf[LogicalPlan], indexes)
+    }.toMap
+  }
+
+  def apply(plan: LogicalPlan, allIndexes: Seq[IndexLogEntry]): PlanToIndexesMap = {
+    val planToIndexes = initializePlanToIndexes(plan, allIndexes)
+    planToIndexes.flatMap {
+      case (node, candidateIndexes) =>
+        Some(node, sourceFilters.foldLeft(candidateIndexes) { (indexes, filter) =>
+          filter(node, indexes)
+        }).filter(_._2.nonEmpty)
+    }
+  }
+}
+
+/**
+ * Apply Hyperspace indexes based on the score of each index application.
+ */
+class ScoreBasedIndexApplication {
+  // TODO: FilterIndexRule :: JoinIndexRule :: Nil
+  private val rules = NoOpRule :: Nil
+
+  // Map for memoization. The key is the logical plan before applying [[HyperspaceRule]]s
+  // and its value is a pair of best transformed plan and its score.
+  private val scoreMap: mutable.HashMap[LogicalPlan, (LogicalPlan, Int)] = mutable.HashMap()
+
+  private def recApply(plan: LogicalPlan, indexes: PlanToIndexesMap): (LogicalPlan, Int) = {
+    // If pre-calculated value exists, return it.
+    scoreMap.get(plan).foreach(res => return res)
+
+    val optResult = (plan, 0)
+    // TODO apply indexes recursively.
+
+    scoreMap.put(plan, optResult)
+    optResult
+  }
+
+  def apply(plan: LogicalPlan, indexes: PlanToIndexesMap): LogicalPlan = {
+    recApply(plan, indexes)._1
+  }
+}
+
+/**
+ * Transform the given plan to use Hyperspace indexes.
+ */
+object ApplyHyperspace
+    extends Rule[LogicalPlan]
+    with Logging
+    with HyperspaceEventLogging
+    with ActiveSparkSession {
+
+  type PlanToIndexesMap = Map[LogicalPlan, Seq[IndexLogEntry]]
```

Comment:
I am thinking whether we should define the following:
```scala
type PlanToCandidateIndexesMap = Map[LogicalPlan, Seq[IndexLogEntry]]
type PlanToSelectedIndexMap = Map[LogicalPlan, IndexLogEntry]
```
, and clearly define what candidate and selected index mean here. WDYT?

Reply:
yep it looks better & consistent 👍